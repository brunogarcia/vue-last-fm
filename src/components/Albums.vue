<template>
  <v-row>
    <v-col
      cols="12"
      lg="8"
      offset-lg="2"
    >
      <!--Album List-->
      <v-card>
        <v-container fluid>
          <v-card-text>
            <div>Genre</div>
            <p class="display-1 text--primary">
              {{ tag }}
            </p>
          </v-card-text>
          <v-row>
            <v-col
              v-for="(album, index) in albums"
              :key="index"
              class="d-flex child-flex"
              cols="12"
              sm="6"
              lg="4"
            >
              <!--Album Image-->
              <v-hover>
                <template v-slot:default="{ hover }">
                  <v-card
                    flat
                    tile
                    class="d-flex"
                  >
                    <!--Image-->
                    <v-img
                      :src="getExtraLargeImage(album.image)"
                      :lazy-src="getSmallImage(album.image)"
                      aspect-ratio="1"
                      class="grey lighten-2"
                    >
                      <!--Overlay-->
                      <v-fade-transition>
                        <v-overlay
                          v-if="hover"
                          absolute
                          color="#036358"
                        >
                          <v-btn :to="getAlbumLink(album)">
                            See more info
                          </v-btn>
                        </v-overlay>
                      </v-fade-transition>

                      <!--Image placeholder-->
                      <template v-slot:placeholder>
                        <v-row
                          class="fill-height ma-0"
                          align="center"
                          justify="center"
                        >
                          <v-progress-circular
                            indeterminate
                            color="grey lighten-5"
                          />
                        </v-row>
                      </template>
                    </v-img>
                  </v-card>
                </template>
              </v-hover>
              <!--end Album Image-->
            </v-col>
          </v-row>
        </v-container>
      </v-card>
      <!--end Album List-->
    </v-col>
  </v-row>
</template>

<script>
import { mapGetters } from 'vuex';
import getImage from '@/utils/getImage';
import constants from '@/utils/constants';

const { SMALL, EXTRA_LARGE } = constants.IMAGES;

export default {
  name: 'AlbumsComponent',

  computed: {
    ...mapGetters({
      tag: 'tags/tag',
      albums: 'albums/albums',
    }),
  },

  methods: {
    /**
     * Get album link
     *
     * @param {Object} album - The album data
     * @returns {String} - The album link
     */
    getAlbumLink(album) {
      const { name: albumName, artist } = album;
      const { name: artistName } = artist;

      return `/album/${artistName}/${albumName}`;
    },

    /**
     * Get small image
     *
     * @param {Array<Object} images - The image list
     * @returns {String} - The album small image
     */
    getSmallImage(images) {
      return getImage(images, SMALL);
    },

    /**
     * Get extra-large image
     *
     * @param {Array<Object} images - The image list
     * @returns {String}- The album extra-large image
     */
    getExtraLargeImage(images) {
      return getImage(images, EXTRA_LARGE);
    },
  },
};
</script>
