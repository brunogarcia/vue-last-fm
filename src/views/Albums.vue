<template>
  <v-container>
    <AlbumsComponent v-if="!loading" />
  </v-container>
</template>

<script>
import { mapActions } from 'vuex';
import AlbumsComponent from '@/components/Albums.vue';

export default {
  name: 'AlbumsView',

  components: {
    AlbumsComponent,
  },

  data: () => ({
    loading: true,
  }),

  created() {
    this.loadData();
  },
  methods: {
    ...mapActions({
      getAlbums: 'albums/getAlbums',
    }),

    async loadData() {
      try {
        await this.getAlbums();
      } catch (error) {
        // show snackbar
      } finally {
        this.loading = false;
      }
    },
  },
};
</script>
