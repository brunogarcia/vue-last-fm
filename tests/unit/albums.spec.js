import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vue from 'vue';
import Vuex from 'vuex';
import Vuetify from 'vuetify';
import Albums from '@/components/Albums.vue';

const localVue = createLocalVue();

Vue.use(Vuetify);

localVue.use(Vuex);

describe('Albums.vue', () => {
  let getters;
  let store;
  let vuetify;

  beforeEach(() => {
    vuetify = new Vuetify();
    getters = {
      albums: () => ({
        name: 'The Perfect Red Velvet - The 2nd Album Repackage',
        mbid: '',
        url: 'https://www.last.fm/music/Red+Velvet',
        artist: {
          name: 'Red Velvet',
          mbid: '4f0cb3b7-6c06-4317-ae35-ddf3106a17ee',
          url: 'https://www.last.fm/music/Red+Velvet',
        },
        image: [
          {
            '#text': 'https://lastfm.freetls.fastly.net/i/u/34s/d31c361f1d65a46ed1d6aeaa99a23b9a.png',
            size: 'small',
          },
          {
            '#text': 'https://lastfm.freetls.fastly.net/i/u/64s/d31c361f1d65a46ed1d6aeaa99a23b9a.png',
            size: 'medium',
          },
          {
            '#text': 'https://lastfm.freetls.fastly.net/i/u/174s/d31c361f1d65a46ed1d6aeaa99a23b9a.png',
            size: 'large',
          },
          {
            '#text': 'https://lastfm.freetls.fastly.net/i/u/300x300/d31c361f1d65a46ed1d6aeaa99a23b9a.png',
            size: 'extralarge',
          },
        ],
        '@attr': {
          rank: '1',
        },
      }),
    };

    store = new Vuex.Store({
      modules: {
        albums: {
          namespaced: true,
          getters,
        },
      },
    });
  });

  it('Renders correctly', () => {
    const wrapper = shallowMount(Albums, {
      store,
      localVue,
      vuetify,
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
