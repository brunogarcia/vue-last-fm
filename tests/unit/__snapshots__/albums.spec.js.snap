// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Albums.vue Renders correctly 1`] = `
<v-row-stub
  tag="div"
>
  <v-col-stub
    cols="12"
    lg="8"
    offsetlg="2"
    tag="div"
  >
    <v-card-stub
      loaderheight="4"
      tag="div"
    >
      <v-container-stub
        fluid="true"
        tag="div"
      >
         
        <v-row-stub
          tag="div"
        >
          <v-col-stub
            class="d-flex child-flex"
            cols="12"
            lg="4"
            sm="6"
            tag="div"
          >
            <v-hover-stub
              closedelay="0"
              opendelay="0"
            />
          </v-col-stub>
          <v-col-stub
            class="d-flex child-flex"
            cols="12"
            lg="4"
            sm="6"
            tag="div"
          >
            <v-hover-stub
              closedelay="0"
              opendelay="0"
            />
          </v-col-stub>
          <v-col-stub
            class="d-flex child-flex"
            cols="12"
            lg="4"
            sm="6"
            tag="div"
          >
            <v-hover-stub
              closedelay="0"
              opendelay="0"
            />
          </v-col-stub>
          <v-col-stub
            class="d-flex child-flex"
            cols="12"
            lg="4"
            sm="6"
            tag="div"
          >
            <v-hover-stub
              closedelay="0"
              opendelay="0"
            />
          </v-col-stub>
          <v-col-stub
            class="d-flex child-flex"
            cols="12"
            lg="4"
            sm="6"
            tag="div"
          >
            <v-hover-stub
              closedelay="0"
              opendelay="0"
            />
          </v-col-stub>
          <v-col-stub
            class="d-flex child-flex"
            cols="12"
            lg="4"
            sm="6"
            tag="div"
          >
            <v-hover-stub
              closedelay="0"
              opendelay="0"
            />
          </v-col-stub>
        </v-row-stub>
      </v-container-stub>
    </v-card-stub>
  </v-col-stub>
</v-row-stub>
`;
